File: src/utils/responses.py
"""
Utilidades para generar respuestas HTTP estandarizadas
"""
import json
from typing import Any, Dict, Optional


def success_response(
    message: str,
    data: Optional[Any] = None,
    status_code: int = 200
) -> Dict[str, Any]:
    """
    Genera una respuesta exitosa estandarizada
    
    Args:
        message: Mensaje descriptivo del éxito
        data: Datos a retornar (opcional)
        status_code: Código HTTP (default: 200)
    
    Returns:
        Dict con statusCode, headers y body
    """
    body = {
        "success": True,
        "message": message
    }
    
    if data is not None:
        body["data"] = data
    
    return {
        "statusCode": status_code,
        "headers": {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": True,
        },
        "body": json.dumps(body, ensure_ascii=False, default=str)
    }


def error_response(
    message: str,
    status_code: int = 400,
    error_type: Optional[str] = None,
    details: Optional[Dict] = None
) -> Dict[str, Any]:
    """
    Genera una respuesta de error estandarizada
    
    Args:
        message: Mensaje descriptivo del error
        status_code: Código HTTP de error
        error_type: Tipo de error (opcional)
        details: Detalles adicionales del error (opcional)
    
    Returns:
        Dict con statusCode, headers y body
    """
    body = {
        "success": False,
        "message": message
    }
    
    if error_type:
        body["error"] = error_type
    
    if details:
        body["details"] = details
    
    return {
        "statusCode": status_code,
        "headers": {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": True,
        },
        "body": json.dumps(body, ensure_ascii=False, default=str)
    }


def bad_request(message: str = "Solicitud inválida", details: Optional[Dict] = None) -> Dict[str, Any]:
    """Respuesta 400 - Bad Request"""
    return error_response(message, 400, "BAD_REQUEST", details)


def unauthorized(message: str = "No autorizado") -> Dict[str, Any]:
    """Respuesta 401 - Unauthorized"""
    return error_response(message, 401, "UNAUTHORIZED")


def forbidden(message: str = "Acceso denegado") -> Dict[str, Any]:
    """Respuesta 403 - Forbidden"""
    return error_response(message, 403, "FORBIDDEN")


def not_found(message: str = "Recurso no encontrado") -> Dict[str, Any]:
    """Respuesta 404 - Not Found"""
    return error_response(message, 404, "NOT_FOUND")


def conflict(message: str = "Conflicto con el estado actual", details: Optional[Dict] = None) -> Dict[str, Any]:
    """Respuesta 409 - Conflict"""
    return error_response(message, 409, "CONFLICT", details)


def internal_error(message: str = "Error interno del servidor") -> Dict[str, Any]:
    """Respuesta 500 - Internal Server Error"""
    return error_response(message, 500, "INTERNAL_ERROR")


def created(message: str, data: Optional[Any] = None) -> Dict[str, Any]:
    """Respuesta 201 - Created"""
    return success_response(message, data, 201)


def no_content() -> Dict[str, Any]:
    """Respuesta 204 - No Content"""
    return {
        "statusCode": 204,
        "headers": {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": True,
        },
        "body": ""
    }

File: src/utils/responses.js
/**
 * Utilidades para generar respuestas HTTP estandarizadas
 */

/**
 * Genera una respuesta exitosa estandarizada
 * 
 * @param {string} message - Mensaje descriptivo del éxito
 * @param {*} [data=null] - Datos a retornar (opcional)
 * @param {number} [statusCode=200] - Código HTTP (default: 200)
 * @returns {Object} Respuesta con statusCode, headers y body
 */
export function successResponse(message, data = null, statusCode = 200) {
  const body = {
    success: true,
    message
  };

  if (data !== null) {
    body.data = data;
  }

  return {
    statusCode,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Credentials': true
    },
    body: JSON.stringify(body)
  };
}

/**
 * Genera una respuesta de error estandarizada
 * 
 * @param {string} message - Mensaje descriptivo del error
 * @param {number} [statusCode=400] - Código HTTP de error
 * @param {string} [errorType=null] - Tipo de error (opcional)
 * @param {Object} [details=null] - Detalles adicionales del error (opcional)
 * @returns {Object} Respuesta con statusCode, headers y body
 */
export function errorResponse(message, statusCode = 400, errorType = null, details = null) {
  const body = {
    success: false,
    message
  };

  if (errorType) {
    body.error = errorType;
  }

  if (details) {
    body.details = details;
  }

  return {
    statusCode,
    headers: {
      'Content-Type': 'application/json',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Credentials': true
    },
    body: JSON.stringify(body)
  };
}

/**
 * Respuesta 400 - Bad Request
 * @param {string} [message='Solicitud inválida'] - Mensaje de error
 * @param {Object} [details=null] - Detalles del error
 * @returns {Object} Respuesta HTTP
 */
export function badRequest(message = 'Solicitud inválida', details = null) {
  return errorResponse(message, 400, 'BAD_REQUEST', details);
}

/**
 * Respuesta 401 - Unauthorized
 * @param {string} [message='No autorizado'] - Mensaje de error
 * @returns {Object} Respuesta HTTP
 */
export function unauthorized(message = 'No autorizado') {
  return errorResponse(message, 401, 'UNAUTHORIZED');
}

/**
 * Respuesta 403 - Forbidden
 * @param {string} [message='Acceso denegado'] - Mensaje de error
 * @returns {Object} Respuesta HTTP
 */
export function forbidden(message = 'Acceso denegado') {
  return errorResponse(message, 403, 'FORBIDDEN');
}

/**
 * Respuesta 404 - Not Found
 * @param {string} [message='Recurso no encontrado'] - Mensaje de error
 * @returns {Object} Respuesta HTTP
 */
export function notFound(message = 'Recurso no encontrado') {
  return errorResponse(message, 404, 'NOT_FOUND');
}

/**
 * Respuesta 409 - Conflict
 * @param {string} [message='Conflicto con el estado actual'] - Mensaje de error
 * @param {Object} [details=null] - Detalles del conflicto
 * @returns {Object} Respuesta HTTP
 */
export function conflict(message = 'Conflicto con el estado actual', details = null) {
  return errorResponse(message, 409, 'CONFLICT', details);
}

/**
 * Respuesta 500 - Internal Server Error
 * @param {string} [message='Error interno del servidor'] - Mensaje de error
 * @returns {Object} Respuesta HTTP
 */
export function internalError(message = 'Error interno del servidor') {
  return errorResponse(message, 500, 'INTERNAL_ERROR');
}

/**
 * Respuesta 201 - Created
 * @param {string} message - Mensaje de éxito
 * @param {*} [data=null] - Datos creados
 * @returns {Object} Respuesta HTTP
 */
export function created(message, data = null) {
  return successResponse(message, data, 201);
}

/**
 * Respuesta 204 - No Content
 * @returns {Object} Respuesta HTTP
 */
export function noContent() {
  return {
    statusCode: 204,
    headers: {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Credentials': true
    },
    body: ''
  };
}

export {
  successResponse,
  errorResponse,
  badRequest,
  unauthorized,
  forbidden,
  notFound,
  conflict,
  internalError,
  created,
  noContent
};